src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -150,6 +173,10 @@ public class CollectionModelBinder<TElement> : IModelBinder
                 {
                     didBind = true;
                     boundValue = result.Model;
+                    if (result.ValidationNode != null)
+                    {
+                        validationNode.ChildNodes.Add(result.ValidationNode);
+                    }
```

Comment:
- Gap if the lower-level binder doesn't create a `ModelValidationNode` but does return a `ModelBindingResult`.  Earlier code unconditionally created a `ModelValidationNode` for each element that was successfully bound (or perhaps for all elements that seem to exist).
- Loops are also incorrect in that they equate `result != null` with success.  Should check `result.IsModelSet` for each element.